### PR TITLE
Add rename popup for extract constant, extract to local variable and extract to function code actions

### LINKED ISSUE #37479: Add rename support for extract refactoring code actions
**Description:**
Add rename support to `extract to constant` and `extract to local variable` code actions.

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/36568
https://github.com/ballerina-platform/ballerina-lang/issues/36569

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
The purpose of this PR is to add rename popup for extract to constant, extract to local variable and extract to function code actions.

Fixes #37479, #38091

Plugin PR [https://github.com/wso2-enterprise/ballerina-plugin-vscode/pull/801](https://github.com/wso2-enterprise/ballerina-plugin-vscode/pull/801)

## Approach
In the create variable code action we have already provided rename popup using the rename support capability. In that case we used the character based position details to determine the start position of the newly created variable. (#37135)

With this approach one limitation was when test this on a windows environment, it gives different position due to the windows detect new lines using `\r\n`.  Where on a linux environment it is represent with `\n`. 

In order to overcome this issue we are using a new capability named positional rename support to provide the rename popup for create variable code action, extract to constant code action, extract to local variable code action and extract to function code action. With this approach we are using line and character based position details.

When the plugin have both the capabilities the language server will prioritise giving the position details using line and character.

## Samples
https://user-images.githubusercontent.com/46857198/201729338-08c8e240-2ae9-4677-af8e-2aa3d6f31e78.mov

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/capability/InitializationOptions.java [hunk 1]
```diff
@@ -57,4 +62,10 @@ public interface InitializationOptions {
      * @return True if supported, false otherwise
      */
     boolean isRefactorRenameSupported();
+
+    /**
+     * Returns if the client supports {@link org.eclipse.lsp4j.Position} based rename.
+     * @return True if supported, false otherwise
+     */
+    boolean isPositionalRefactoredRenameSupported();
```
Comment (TharushiJay) [inline language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/capability/InitializationOptions.java:70]:
```suggestion
    boolean isPositionalRefactorRenameSupported();
```

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableWithPositionalRenameSupportCapabilityTest.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableWithPositionalRenameSupportCapabilityTest.java:2]:
```suggestion
 *  Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableWithPositionalRenameSupportCapabilityTest.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+import org.ballerinalang.langserver.util.FileUtils;
+import org.ballerinalang.langserver.util.TestUtil;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+/**
+ * Test Cases for CodeActions.
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableWithPositionalRenameSupportCapabilityTest.java:31]: Let's make this description more specific

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableWithPositionalRenameSupportCapabilityTest.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    @Override
+    protected void setupLanguageServer(TestUtil.LanguageServerBuilder builder) {
+        builder.withInitOption(InitializationOptions.KEY_POSITIONAL_RENAME_SUPPORT, true);
+    }
+
+    @Override
+    protected Path getConfigJsonPath(String configFilePath) {
+        return FileUtils.RES_DIR.resolve("codeaction")
+                .resolve(getResourceDir())
+                .resolve("config-ren-pos-cap")
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableWithPositionalRenameSupportCapabilityTest.java:46]: Shall we avoid using shorthands like this? It's a little difficult to make sense of what it means, at first sight. Let's use something like `config-rename-positional`/`config-rename-positional-capability` instead.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/ImportsAcceptor.java [hunk 5]
```diff
@@ -127,7 +127,7 @@ private TextEdit createImportTextEdit(String pkgName) {
         int endCol = 0;
         int endLine = lastImport.isEmpty() ? 0 : lastImport.get().location().lineRange().endLine().line();
 
-        String editText = "import " + pkgName + ";\n";
+        String editText = String.format("import %s;%n", pkgName);
```
Comment (mindula) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/ImportsAcceptor.java:130]: Shall we use `ItemResolverConstants.IMPORT`?

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AbstractCodeActionTest.java [hunk 6]
```diff
@@ -394,23 +394,54 @@ protected boolean validateAndModifyArguments(JsonObject actualCommand,
...
+                    newArgs.add(actualRenamePosition);
+
+                    //Replace the args of the actual command to update the test config
+                    actualCommand.add("arguments", newArgs);
+                }
+            }
+            return false;
+        }
+
+        if ("ballerina.action.positional.rename".equals(actualCommand.get("command").getAsString())) {
```
Comment (mindula) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AbstractCodeActionTest.java:429]: Let's move `ballerina.action.positional.rename` to a constant

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableTest.java [hunk 7]
```diff
@@ -47,6 +61,52 @@ public void negativeTest(String config) throws IOException, WorkspaceDocumentExc
...
 
+    @Override
+    protected boolean validateAndModifyArguments(JsonObject actualCommand,
+                                                 JsonArray actualArgs,
+                                                 JsonArray expArgs,
+                                                 Path sourceRoot,
+                                                 Path sourcePath,
+                                                 List<TextEdit> actualEdits,
+                                                 TestConfig testConfig) {
+        if ("ballerina.action.rename".equals(actualCommand.get("command").getAsString())) {
```
Comment (mindula) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateVariableTest.java:72]: Lets use a constant for `ballerina.action.rename`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java [hunk 8]
```diff
@@ -145,6 +163,24 @@ private boolean isExpressionNode(Node node) {
...
+        LSClientCapabilities lsClientCapabilities = context.languageServercontext().get(LSClientCapabilities.class);
+        if (lsClientCapabilities.getInitializationOptions().isPositionalRefactorRenameSupported()) {
+            codeAction.setCommand(new Command(
+                    CommandConstants.RENAME_COMMAND_TITLE_FOR_CONSTANT, CommandConstants.POSITIONAL_RENAME_COMMAND,
+                    List.of(context.fileUri(), getRenamePosition(replacePosition, addNewLineAtStart))));
+        }
+    }
+
+    private Position getRenamePosition(Position replacePosition, boolean addNewLineAtStart) {
+        int line = replacePosition.getLine() + 1;
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java:177]: Can you add a comment explaining why we have to add +1 to the line here

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToLocalVarCodeAction.java [hunk 9]
```diff
@@ -134,14 +137,26 @@ public List<CodeAction> getCodeActions(CodeActionContext context,
...
+    }
 
-        return Collections.singletonList(CodeActionUtil.createCodeAction(CommandConstants.EXTRACT_TO_VARIABLE, 
-                List.of(varDeclEdit, replaceEdit), context.fileUri(), CodeActionKind.RefactorExtract));
+    private void addRenamePopup(CodeActionContext context, CodeAction codeAction, Position position) {
+        LSClientCapabilities lsClientCapabilities = context.languageServercontext().get(LSClientCapabilities.class);
+        if (lsClientCapabilities.getInitializationOptions().isPositionalRefactorRenameSupported()) {
+            codeAction.setCommand(new Command(
+                    CommandConstants.RENAME_COMMAND_TITLE_FOR_VARIABLE, CommandConstants.POSITIONAL_RENAME_COMMAND,
+                    List.of(context.fileUri(), new Position(position.getLine() + 1, position.getCharacter()))));
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToLocalVarCodeAction.java:158]: ditto

**Review by malinthar — CHANGES_REQUESTED** [review]

**Comment by malinthar** [comment]
Getting the `element can't be renamed`  message for the following scenario. The rename position may be wrong in this case.
rename_pop_up

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java [hunk 10]
```diff
@@ -145,6 +163,24 @@ private boolean isExpressionNode(Node node) {
         return node instanceof StatementNode || node instanceof ModuleMemberDeclarationNode;
     }
 
+    private void addRenamePopup(CodeActionContext context, CodeAction codeAction, Position replacePosition,
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java:166]: Shall we make this method a util method as this is used in multiple code actions

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java [hunk 11]
```diff
@@ -106,24 +106,42 @@ public List<CodeAction> getCodeActions(CodeActionContext context,
...
+            CodeAction codeAction = CodeActionUtil.createCodeAction(CommandConstants.EXTRACT_TO_CONSTANT,
+                    textEdits, context.fileUri(), CodeActionKind.RefactorExtract);
+            addRenamePopup(context, codeAction, textEdits.get(1).getRange().getStart(), addNewLineAtStart);
+            return Collections.singletonList(codeAction);
         }
 
         // Selection is a position
         List<Node> nodeList = getPossibleExpressionNodes(node, nodeValidator);
-        LinkedHashMap<String, List<TextEdit>> textEditMap = new LinkedHashMap<>();
+        if (nodeList.size() == 1) {
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java:125]: We cannot move. If we are moving the condition we have to move this `List<Node> nodeList = getPossibleExpressionNodes(node, nodeValidator);` statement asswell.

**Comment by LakshanWeerasinghe** [comment]
> Getting the `element can't be renamed` message for the following scenario. The rename position may be wrong in this case. rename_pop_up [ rename_pop_up ](https://user-images.githubusercontent.com/35211477/211767746-8bbc3fef-ca8d-487f-b3bf-f25fcbb06d5e.gif) [ ](https://user-images.githubusercontent.com/35211477/211767746-8bbc3fef-ca8d-487f-b3bf-f25fcbb06d5e.gif)

Fixed this. Please use the updated PR for plugin as well.

**Review by IMS94 — APPROVED** [review]

**Review by malinthar — APPROVED** [review]
LGTM!

**Pull request merged**